debezium-server/debezium-server-ybexporter/src/main/java/io/debezium/server/ybexporter/StreamingWriterJson.java

```diff
@@ -5,87 +5,137 @@
+ * as the file size may become too large. Therefore, we break it into smaller QueueSegments
+ * and rotate them as soon as we reach a size threshold.
+ * If shutdown abruptly, this class is capable of resuming by retrieving the latest queue
+ * segment that was written to, and continues writing from that segment.
+ */
 public class StreamingWriterJson implements RecordWriter {
-    private static final Logger LOGGER = LoggerFactory.getLogger(TableSnapshotWriterCSV.class);
-    private static final String QUEUE_FILE_NAME = "queue.json";
+    private static final Logger LOGGER = LoggerFactory.getLogger(StreamingWriterJson.class);
+    private static final String QUEUE_FILE_NAME = "queue";
+    private static final String QUEUE_FILE_EXTENSION = "ndjson";
+    private static final String QUEUE_FILE_DIR = "cdc";
+    private static final long QUEUE_SEGMENT_MAX_BYTES = 200 * 1000 * 1000; // 200 MB
     private String dataDir;
-    private BufferedWriter writer;
-    private ObjectWriter ow;
-    private FileOutputStream fos;
-    private FileDescriptor fd;
+    private QueueSegment currentQueueSegment;
+    private long currentQueueSegmentIndex = 0;
+
 
     public StreamingWriterJson(String datadirStr) {
         dataDir = datadirStr;
 
-        var fileName = String.format("%s/%s", dataDir, QUEUE_FILE_NAME);
-        try {
-            fos = new FileOutputStream(fileName, true);
-            fd = fos.getFD();
-            var f = new FileWriter(fd);
-            writer = new BufferedWriter(f);
-
+        // mkdir cdc
+        File queueDir = new File(String.format("%s/%s", dataDir, QUEUE_FILE_DIR));
+        if (!queueDir.exists()){
+            boolean dirCreated = queueDir.mkdir();
+            if (!dirCreated){
+                throw new RuntimeException("failed to create dir for cdc");
+            }
         }
-        catch (IOException e) {
-            throw new RuntimeException(e);
+        recoverStateFromDisk();
+        if (currentQueueSegment == null){
+            createNewQueueSegment();
         }
-        ow = new ObjectMapper().writer().withDefaultPrettyPrinter();
     }
 
-    @Override
-    public void writeRecord(Record r) {
+    private void createNewQueueSegment(){
+        currentQueueSegment = new QueueSegment(getFilePathWithIndex(currentQueueSegmentIndex));
+    }
+
+    private void recoverStateFromDisk(){
+        recoverLatestQueueSegment();
+    }
+
+    /**
+     * This function reads the /cdc dir for all the queue segment files.
+     * Then it retrieves the index number from the paths, and then finds
+     * the max index - which is the latest queue segment that was written to.
+     * If no files are found, we just return.
+     */
+    private void recoverLatestQueueSegment(){
+        // read dir to find all queue files
+        Path queueDirPath = Path.of(dataDir, QUEUE_FILE_DIR);
+        String searchGlob = String.format("%s.*.%s", QUEUE_FILE_NAME, QUEUE_FILE_EXTENSION);
```

Comment:
Instead of `queue` use `segment`.
Instead of `cdc` use `queue`.

In other words, a "queue" is represented as a directory containing bunch of "queue segments".

In that case, the `StreamingWriterJson` is in fact `Queue` class ;-) (or `EventQueue` to be more precise).

Reply:
Indeed! :)